support/scripts/multiboot2.py

```diff
@@ -0,0 +1,245 @@
+
+            f.seek(self.e_shoff)
+            self.elf64_shdrs = []
+            for _ in range(self.e_shnum):
+                self.elf64_shdrs.append(f.read(ehdr64_e_shentsize))
+
+    # We use this function to adjust offsets w.r.t. the prepended ELF64 header,
+    # the Multiboot header and the Program headers.
+    def prpnd_off(self, barray, mb2_header_end_off):
+        sz = len(barray)
+        old_off = int.from_bytes(barray, self.endian)
+
+        new_off = mb2_header_end_off + ELF64_PHDR_LEN * self.e_phnum + old_off
+        if sz < (new_off.bit_length() + 7) // 8:
+            raise Exception("New size exceeds initial byte-length.")
+
+        return new_off.to_bytes(sz, self.endian)
+
+    def get_ehdr(self, mb2_header_end_off):
+        elf64_ehdr = bytearray(ELF64_EHDR_LEN)
+        elf64_ehdr[0:ELF64_EHDR_LEN] = self.elf64_ehdr[0:ELF64_EHDR_LEN]
+
+        # The offset to the ELF64 Program/Sections Headers are at
+        # the end of the Multiboot Header and the end of the file respectively,
+        # so that we do not mess up the Multiboot2 header, which must exist
+        # in the first 8192 bytes. Although the specification does not enforce
+        # this, GRUB, for whatever reason, wants the Program Headers to also be
+        # be placed in the first 8192 bytes (see commit 9a5c1ad).
+        # Therefore the final image will have, in this order:
+        # ELF64 header
+        # Multiboot2 header
+        # ELF64 Program headers
+        # Original ELF64 binary
+        # ELF64 Section headers (these are too many so place them at the end)
+        elf64_phoff = mb2_header_end_off
+        elf64_shoff = mb2_header_end_off + ELF64_PHDR_LEN * self.e_phnum + self.file_sz
+
+        elf64_ehdr[32:40] = elf64_phoff.to_bytes(8, self.endian)  # e_phoff
+        elf64_ehdr[40:48] = elf64_shoff.to_bytes(8, self.endian)  # e_shoff
+
+        return elf64_ehdr
+
+    def get_phdrs(self, mb2_header_end_off):
+        def update_phdr(phdr):
+            new_phdr = bytearray(ELF64_PHDR_LEN)
+
+            new_phdr[0:ELF64_PHDR_LEN] = phdr[0:ELF64_PHDR_LEN]  # keep the rest
+            new_phdr[8:16] = self.prpnd_off(phdr[8:16], mb2_header_end_off)  # p_offset
+
+            return new_phdr
+
+        return [update_phdr(p) for p in self.elf64_phdrs]
+
+    def get_shdrs(self, mb2_header_end_off):
+        def update_shdr(shdr):
+            new_shdr = bytearray(ELF64_SHDR_LEN)
+
+            new_shdr[0:ELF64_SHDR_LEN] = shdr[0:ELF64_SHDR_LEN]  # keep the rest
+            new_shdr[24:32] = self.prpnd_off(
+                shdr[24:32], mb2_header_end_off
+            )  # sh_offset
+
+            return new_shdr
+
+        return [update_shdr(s) for s in self.elf64_shdrs]
+
+
+def add_tags(tags_opt):
+    tags_hdr = b""
+    for tag_type in tags_opt:
+        if (
+            tag_type == MULTIBOOT_HEADER_TAG_END
+            or tag_type == MULTIBOOT_HEADER_TAG_MODULE_ALIGN
+        ):
+            size = 8
```

Comment:
what is this `size` and does it need to be assigned in each iteration? seems pretty constant to me.

Reply:
The `size` field is the total size the tag will have with all of its fields. In the case of the implemented tags, it is indeed equal to 8. However, there are other tags whose sizes can differ, as stated in the specification  (e.g. `MULTIBOOT_HEADER_TAG_RELOCATABLE`  would have `size` equal to 24). Should I use a macro for now and change it up once more tags are added?

Reply:
Understood. There can only be one instance of a tag type correct? I am thinking maybe we should just create an individual function for each. E.g.
```python
def get_multiboot2_hdr():
    mb2_magic = MULTIBOOT_HEADER_MAGIC
    mb2_architecture = MULTIBOOT_ARCHITECTURE_I386

    tags_hdr = multiboot2_tag_align()
    tags_hdr += multiboot2_tag_end()
    tags_size = len(tags_hdr)
    ....
 ```
 instead of the `add_tags` function. So when we would add a certain tag in the future we would create a `multiboot2_tag_<tag_name>(<potential_args>)`, where potential_args would be possible values that may change the behavior of the tag, e.g. `multiboot2_tag_reloc(min_addr, max_addr, align, preference)`.
 
Sorry, when I first made the suggestion, I thought we could like make a very generic class structure or something to generically instantiate tags, but now that I see them properly in the spec they can differ quite a log, some not even having the `flags` field... it may not be worth the future maintenance effort... I think it would be easier to just have a function that returns a tag for type of tag.
What do you think?

Reply:
Yes, it would be better and cleaner this way. It could get complicated/ugly if complex tags are added down the road. I will update the implementation :smiley: